src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -198,23 +196,41 @@ bool InstanceKlass::has_nest_member(InstanceKlass* k, TRAPS) const {
       if (name == k->name()) {
         log_trace(class, nestmates)("- Found it at nest_members[%d] => cp[%d]", i, cp_index);
 
-        // Names match so check actual klass. This may trigger class loading if
-        // it doesn't match though that should be impossible as it means one classloader
-        // has defined two different classes with the same name! A compiler thread won't be
-        // able to perform that loading but we can't exclude the compiler threads from
-        // executing this logic. But it should actually be impossible to trigger loading here.
-        Klass* k2 = _constants->klass_at(cp_index, THREAD);
-        assert(!HAS_PENDING_EXCEPTION || PENDING_EXCEPTION->is_a(vmClasses::VirtualMachineError_klass()),
-               "Exceptions should not be possible here");
-        if (k2 == k) {
-          log_trace(class, nestmates)("- class is listed as a nest member");
-          return true;
-        }
-        else {
-          // same name but different klass!
-          log_trace(class, nestmates)(" - klass comparison failed!");
-          // can't have two names the same, so we're done
-          return false;
+        // Names match so check actual klass, as we need identity between classes not equivalence.
+        // It should not be possible for constant-pool resolution to trigger actual class-loading,
+        // as that would imply one classloader has defined two classes with the same name. But it
+        // is possible that the resolution process could require protection domain validation, which
+        // requires executing Java code, and can lead to exceptions (like OOME or StackOverflow).
+        // So we have to skip constant-pool resolution in that case.
+        if (can_resolve) {
+          Klass* k2 = _constants->klass_at(cp_index, THREAD);
```

Comment:
why is this not just CHECK_NULL?

Reply:
Logging. Plus the check for only VME which probably has to go anyway.